The notebook of Preprocessor class (#60)
Saransh-cpp/OCRed: examples/Preprocessing_example.md

# Preprocessign an Image

The main objective of the Preprocessing phase is to make as easy as possible for the OCR system to distinguish a character/word from the background. There are several techniques used for Preprocessing but this **Preprocessor** class can do it without you going into details and jargon of traditional methods.

`!pip install -U ocred --no-cache --force-reinstall`

```
# manually preprocessing an image
import cv2

> from scipy import ndimage

Reviewer: See the last comment!
```suggestion
```

from ocred import Preprocessor

> from google.colab.patches import cv2_imshow

Reviewer: I think we can use `cv2` or `matplotlib` to show images in `Jupyter`. (`cv2.imshow` should work and this import can be removed)

Author: I did use `cv2.imshow` initially but it gave an error that this function is not functional anymore in colab, hence I had to import the new function separately and use it.

> preprocessed = Preprocessor("/content/images/CosmosOne.jpg")

Reviewer: Most of the users will be using `Jupyter` locally. You can make this change at the end, when you are done with running this code on colab!
```suggestion
preprocessed = Preprocessor("../images/CosmosOne.jpg")
```

```

```
# scan the image and copy the scanned image

> scanned = preprocessed.scan(inplace=True)
> orig = scanned.copy()
> 
> cv2_imshow(orig)

Reviewer: It might be good to show users the `inplace=False` version and then point to the `inplace=True` version.

`inplace=True` edits the `overriden_image` that is passed to the the function, which is memory-wise efficient (hence time-wise efficient - the time taken by OS to create a new array for the copy of image is saved), but you might not want the program to alter your original image.

`inplace=False` on the other hand creates a copy of the passed image, hence more memory and time usage but does not alter the original image -
https://github.com/Saransh-cpp/OCRed/blob/420aa7a78e75e1ea7963079175260997a23816a4/ocred/preprocessing.py#L84-L89

If an `overriden_image` is not passed, then this is performed with the image passed to the class.

```suggestion
scanned = preprocessed.scan()

cv2_imshow(scanned)
```

```

![image](https://user-images.githubusercontent.com/82112540/184899616-2a418f9b-ad82-4bf4-92b7-98a96126dbd1.png)

```
# remove noise

> noise_free = preprocessed.remove_noise(
>     inplace=True, overriden_image=scanned
> )
> 
> cv2_imshow(noise_free)

Reviewer: Regarding `inplace` (this would make it more clear) -

See how we are setting `inplace=True` and passing in an `overriden_image` named `scanned`. This would edit the actual `scanned` variable, hence if you look at `scanned` after running the function, it would be noise free! This is efficient and this also would not require capturing the value of the returned variable.

On the other hand if you turn `inplace=False`, then `ocred` will generate a copy of `scanned`, remove noise from that copy, and return that copy. This copy can be captured as the return value and the `scanned` variable will not be altered. (not very efficient)

```suggestion
noise_free = preprocessed.remove_noise(overriden_image=scanned)

cv2_imshow(noise_free)
```

```

![image](https://user-images.githubusercontent.com/82112540/184900366-34551949-42f7-4b6c-b448-8a0a6fb05779.png)

```

> # thicken the ink to draw Hough lines better
> thickened = preprocessed.thicken_font(
>     inplace=True, overriden_image=noise_free
> )
> 
> cv2_imshow(thickened)

Reviewer: Ditto!
```suggestion
# thicken the ink to draw Hough lines better
thickened = preprocessed.thicken_font(overriden_image=noise_free)

cv2_imshow(thickened)
```

```

![image](https://user-images.githubusercontent.com/82112540/184900996-2eec35d6-ef00-4fc1-bd8d-34944c793b0e.png)

```

> # calculate the median angle of all the Hough lines
> _, median_angle = preprocessed.rotate(
>     inplace=True, overriden_image=thickened
> )
> 
> cv2_imshow(_)

Reviewer: ```suggestion
# calculate the median angle of all the Hough lines
rotated, median_angle = preprocessed.rotate(overriden_image=thickened)

cv2_imshow(rotated)
```

```

![image](https://user-images.githubusercontent.com/82112540/184901244-741ea02d-06de-4d39-a9a1-ce04c2eabf91.png)

```
# rotate the original scanned image
rotated = ndimage.rotate(orig, median_angle)

cv2_imshow(rotated)
```

![image](https://user-images.githubusercontent.com/82112540/184902055-61c50a47-7894-4d2d-aba7-89acfa4b53f1.png)

```
# remove noise again
final_img = preprocessed.remove_noise(inplace=True, overriden_image=rotated)

cv2_imshow(final_img)
```

> ![image](https://user-images.githubusercontent.com/82112540/184902140-42582eac-e765-44b6-b518-b8aa81ada09c.png)

Reviewer: I think we can remove these, and add the `inplace=True` example.

Notice how we don't need return values now, given that the functions are altering the `pre_img` variable directly, without copying it, making the code efficient!

Would be nice to split this into cells too!
```py
preprocessed = Preprocessor("path/to/image")

# scan the image and copy the scanned image
pre_img = preprocessed.scan(inplace=True)

# plot pre_img

# remove noise
preprocessed.remove_noise(
    inplace=True, overriden_image=pre_img  # the variable `pre_img` is altered 'inplace'
)

# plot pre_img

# thicken the ink to draw Hough lines better
preprocessed.thicken_font(
    inplace=True, overriden_image=pre_img
)

# plot pre_img

# calculate the median angle of all the Hough lines
preprocessed.rotate(
    inplace=True, overriden_image=pre_img
)

# plot pre_img
```

Author: What should I use to plot the image ? Also, I think we need variable to save the changes and then plot the image because although we put `inplace=True` when I tried to plot using `cv2_imshow(pre-img)` the outputs didn't have any changes.
